homeassistant/components/smappee/config_flow.py

```diff
@@ -1,23 +1,178 @@
+                CONF_SERIALNUMBER: serial_number,
+                "title_placeholders": {"name": serial_number},
+            }
+        )
+
+        return await self.async_step_zeroconf_confirm()
+
+    async def async_step_zeroconf_confirm(self, user_input=None):
+        """Confirm zeroconf flow."""
+        errors = {}
+
+        # Check if already configured (cloud)
+        if self.is_cloud_device_already_added():
+            return self.async_abort(reason="already_configured_device")
+
+        if user_input is None:
+            # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+            serialnumber = self.context.get(CONF_SERIALNUMBER)
+            return self.async_show_form(
+                step_id="zeroconf_confirm",
+                description_placeholders={"serialnumber": serialnumber},
+                errors=errors,
+            )
+
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        ip_address = self.context.get(CONF_IP_ADDRESS)
+        serial_number = self.context.get(CONF_SERIALNUMBER)
+
+        # Attempt to make a connection to the local device
+        smappee_api = api.api.SmappeeLocalApi(ip=ip_address)
+        logon = await self.hass.async_add_executor_job(smappee_api.logon)
+        if logon is None:
+            return self.async_abort(reason="connection_error")
+
+        return self.async_create_entry(
+            title=f"{DOMAIN}{serial_number}",
+            data={CONF_IP_ADDRESS: ip_address, CONF_SERIALNUMBER: serial_number},
+        )
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initiated by the user."""
+
+        # If there is a CLOUD entry already, abort a new LOCAL entry
+        if self.is_cloud_device_already_added():
+            return self.async_abort(reason="already_configured_device")
+
+        return await self.async_step_environment()
+
+    async def async_step_environment(self, user_input=None):
+        """Decide environment, cloud or local."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="environment",
+                data_schema=vol.Schema(
+                    {
+                        vol.Required("environment", default=ENV_CLOUD): vol.In(
+                            [ENV_CLOUD, ENV_LOCAL]
+                        )
+                    }
+                ),
+                errors={},
+            )
+
+        # Environment chosen, request additional host information for LOCAL or OAuth2 flow for CLOUD
+        # Ask for host detail
+        if user_input["environment"] == ENV_LOCAL:
+            return await self.async_step_local()
+
+        # Abort cloud option if a LOCAL entry has already been added
+        if (
+            user_input["environment"] == ENV_CLOUD
+            and len(self._async_current_entries()) > 0
+        ):
+            return self.async_abort(reason="already_configured_device")
```

Comment:
Do you want to add a special abort reason here, telling the user to remove existing local entries before configuring the cloud?

Reply:
Added the _already_configured_local_device_ abort reason.